### PR TITLE
Add oEmbed Support

### PR DESCRIPTION
**Pull request by jj0hns0n**
This PR adds oEmbed support to mapstory enabling it to be both a consumer and provider of oEmbed functionality (we only care about being a provider now, but consumer support is there should we need it).

This relies on this library https://github.com/worldcompany/djangoembed/ which very clearly says it is unmaintained, yet the alternate library it mentions does not include oEmbed provider support. I tried to do a manual implementation of oEmbed provider and could not get it to work with Wordpress, so gave up and used this library.

I've tested this with  Wordpress and it works if you add the following lines as a plugin to your wordpress install. The best path forward here is to make an actual mapstory plugin for wordpress that is easy to install. That should be trivial, but requires a bit of discussion on how its to be distributed etc. Beyond that, wordpress would have to whitelist our site in order to make it work directly with hosted sites on wordpress.com. We can also look into embed.ly as an option.

function wpb_oembed_mapstory(){
wp_oembed_add_provider( 'http://mapstory.dev.opengeo.org/maps/*', 'http://mapstory.dev.opengeo.org/oembed/json' );
}
add_action('init','wpb_oembed_mapstory');

One other thing to note is that the maps embed page is not currently setup properly. It displays the entire page headers and all. We need to make this file https://github.com/opengeo/mapstory/blob/master/templates/maps/embed.html look more like this one https://github.com/opengeo/geonode/blob/mapstory/src/GeoNodePy/geonode/templates/maps/embed.html hoping that @ischneider can help me with that as I dont have my head fully wrapped around the way the viewer works.

@cholmes I had this stood up on dev, but I believe it was re-deployed. Once this PR is merged, we can test again.

### CODE REVIEW

**Comment by jj0hns0n** [comment]
Forgot to mention, here is an example of this (for as long as it stays up, or until this gets merged).

http://mapstory.dev.opengeo.org/oembed/json/?url=http%3A%2F%2Fmapstory.dev.opengeo.org%2Fmaps%2F263

and here is that used in a wordpress blog post.

http://ec2-184-72-180-109.compute-1.amazonaws.com/wordpress/

File: oembed_providers.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    regex = settings.SITEURL.replace('/', '\/') + 'maps\/\d+' 
+    print regex
+    provides = True 
+    resource_type = 'rich'
+    
+    def request_resource(self, url, **kwargs):
+        pattern = settings.SITEURL.replace('/', '\/') + 'maps\/\d+'
+        result = re.match(pattern, url)
+        if result is None:
+            return HttpResponse(("Invalid oEmbed URL"), status=400, mimetype="text/plain")
```
Comment (iwillig) [inline oembed_providers.py:21]: Hey Jeff, 

I think we are missing an import here. I get an http response undefined error

File: oembed_providers.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import oembed
```
Comment (iwillig) [inline oembed_providers.py:1]: Hey, I think this is an unused import.

File: oembed_providers.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    
+    def request_resource(self, url, **kwargs):
+        pattern = settings.SITEURL.replace('/', '\/') + 'maps\/\d+'
+        result = re.match(pattern, url)
+        if result is None:
+            return HttpResponse(("Invalid oEmbed URL"), status=400, mimetype="text/plain")
+        else:
+            map_id = int(urlparse(url).path.split('/')[2])
+            map_obj = get_object_or_404(Map, id=map_id)
+            #if not request.user.has_perm('maps.view_map', obj=map_obj):
```
Comment (iwillig) [inline oembed_providers.py:25]: If this is commented out, can we remove it from the PR?

File: oembed_providers.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+from urlparse import urlparse
+from django.shortcuts import get_object_or_404
+from geonode.maps.models import Map
+
+class MapStoryProvider(BaseProvider):
+    regex = settings.SITEURL.replace('/', '\/') + 'maps\/\d+' 
+    print regex
+    provides = True 
+    resource_type = 'rich'
+
```
Comment (iwillig) [inline oembed_providers.py:16]: Extra whitespace on this line.

File: oembed_providers.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        
+        # prepare the dictionary of data to be returned as an oembed resource
+        data = {
+            'type': 'rich', 'provider_name': 'MapStory', 'version': '1.0',
+            'width': maxwidth, 'height': maxheight, 'title': map_obj.title, 'author_name': map_obj.owner.username,
+            'author_url': settings.SITEURL[:-1] + map_obj.owner.get_absolute_url() 
+        }
+        embed_url = settings.SITEURL[:-1] + map_obj.get_absolute_url() + '/embed'
+        data['html'] = '<iframe style="border: none;" height="%s" width="%s" src="%s"></iframe>' % (maxheight, maxwidth, embed_url) 
+
```
Comment (iwillig) [inline oembed_providers.py:38]: Extra whitespace on this line.

File: oembed_providers.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        if result is None:
+            return HttpResponse(("Invalid oEmbed URL"), status=400, mimetype="text/plain")
+        else:
+            map_id = int(urlparse(url).path.split('/')[2])
+            map_obj = get_object_or_404(Map, id=map_id)
+            #if not request.user.has_perm('maps.view_map', obj=map_obj):
+            #    return HttpResponse(("Not Permitted"), status=401, mimetype="text/plain")
+        maxwidth = kwargs.get('maxwidth', None)
+        maxheight = kwargs.get('maxheight', None)
+
```
Comment (iwillig) [inline oembed_providers.py:29]: Extra whitespace on this line.

File: oembed_providers.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
+import oembed
+from oembed.providers import BaseProvider
+from oembed.resources import OEmbedResource
+from oembed.utils import size_to_nearest, scale
```
Comment (iwillig) [inline oembed_providers.py:4]: We don't use scale I don't think.

File: oembed_providers.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+from django.conf import settings
+import re
+from urlparse import urlparse
+from django.shortcuts import get_object_or_404
+from geonode.maps.models import Map
+
```
Comment (iwillig) [inline oembed_providers.py:10]: Two blank lines for a new class def

**Comment by ischneider** [comment]
as for the /embed view, see #384 as the corresponding branch

**Comment by ischneider** [comment]
closed with ed43849b

**Pull request closed without merging**